File: server/src/test_client/AirQualityClient.java
package test_client;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import lombok.SneakyThrows;
import sr.grpc.generated.air_quality.AirQuality.AirQualityInfo;
import sr.grpc.generated.air_quality.AirQuality.AirQualityRequest;
import sr.grpc.generated.air_quality.AirQualityServiceGrpc;
import sr.grpc.generated.air_quality.AirQualityServiceGrpc.AirQualityServiceBlockingStub;
import util.ConsoleColor;

import java.util.Iterator;
import java.util.Scanner;
import java.util.UUID;

import static java.lang.Thread.sleep;
import static util.ColouredPrinter.printlnColoured;

public class AirQualityClient {

    private final ManagedChannel channel;
    private final AirQualityServiceBlockingStub blockingStub;

    public AirQualityClient(String host, int port) {
        channel = ManagedChannelBuilder.forAddress(host, port)
                .usePlaintext()
                .build();

        printlnColoured("Client started", ConsoleColor.GREEN_BOLD);

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            printlnColoured("Shutting down client", ConsoleColor.RED_BOLD);
            channel.shutdown();
        }));

        blockingStub = AirQualityServiceGrpc.newBlockingStub(channel);
    }

    public void start() {
        subscribeOnAirQualityInfo();
        channel.shutdown();
    }

    @SneakyThrows
    private void subscribeOnAirQualityInfo() {
        final String cityName = chooseCityToGetInfoAbout();

        final AirQualityRequest request = AirQualityRequest
                .newBuilder()
                .setCity(cityName)
                .setClientIdentifier(UUID.randomUUID().toString())
                .build();

        Iterator<AirQualityInfo> airQualityInfoIterator;
        while (true) {
            try {
                airQualityInfoIterator = blockingStub.subscribeOnAirQuality(request);

                while (airQualityInfoIterator.hasNext()) {
                    AirQualityInfo response = airQualityInfoIterator.next();
                    printlnColoured(response.toString(), ConsoleColor.YELLOW_BOLD);
                }

            } catch (Exception e) {
                printlnColoured("Connection issues....", ConsoleColor.RED_BOLD_BRIGHT);
                sleep(1000);
            }
        }
    }

    private String chooseCityToGetInfoAbout() {
        printlnColoured("Choose name of the city that you want to subscribe for air quality info: ", ConsoleColor.CYAN_BOLD);
        Scanner scanner = new Scanner(System.in);
        return scanner.nextLine();
    }
}

File: grpc/client_grpc/air_quality_client.py
import threading
import time
import uuid

from printer import *

import grpc

from gen import air_quality_pb2
from gen import air_quality_pb2_grpc


def run():
    city_name = input("Choose name of the city that you want to subscribe for air quality info: ")
    print_coloured(city_name, CVIOLET)

    while True:
        try:
            channel = create_channel()
            print_coloured("Client's connecting...", CGREEN)

            air_quality_service_stub = air_quality_pb2_grpc.AirQualityServiceStub(channel)
            client_identifier = str(uuid.uuid1())
            request = create_request(city_name, client_identifier)

            future_result = air_quality_service_stub.SubscribeOnAirQuality(request)

            for item in future_result:
                print_coloured(str(item), CGREEN)

            print_coloured('Server has ended processing', CGREEN)

        except:
            print_coloured("Connection issues...", CRED)
            time.sleep(1)


def create_channel():
    return grpc.insecure_channel(
        'localhost:50051',
        options=[
            ('grpc.keepalive_time_ms', 10000),
            ('grpc.keepalive_timeout_ms', 5000),
            ('grpc.keepalive_permit_without_calls', True)
        ]
    )


def create_request(city_name, client_identifier):
    return air_quality_pb2.AirQualityRequest(
        city=city_name,
        clientIdentifier=client_identifier)


def get_air_quality_info():
    threading.Thread(
        target=run,
        args="", daemon=True).start()


get_air_quality_info()

while True:
    pass
